src/renderer.ts

```diff
@@ -225,94 +225,97 @@ class Renderer extends EventEmitter<RendererEvents> {
     return gradient
   }
 
-  private renderSingleCanvas(
+  private renderBars(
     channelData: Array<Float32Array | number[]>,
     options: WaveSurferOptions,
-    width: number,
-    start: number,
-    end: number,
-    canvasContainer: HTMLElement,
-    progressContainer: HTMLElement,
+    ctx: CanvasRenderingContext2D,
   ) {
-    const pixelRatio = window.devicePixelRatio || 1
-    const height = options.height || 0
-    const barWidth = options.barWidth != null && !isNaN(options.barWidth) ? options.barWidth * pixelRatio : 1
-    const barGap =
-      options.barGap != null && !isNaN(options.barGap)
-        ? options.barGap * pixelRatio
-        : options.barWidth
-        ? barWidth / 2
-        : 0
-    const barRadius = options.barRadius || 0
-    const barScale = options.barHeight || 1
-
-    const isMono = channelData.length === 1
-    const leftChannel = channelData[0]
-    const rightChannel = isMono ? leftChannel : channelData[1]
-    const useNegative = isMono && rightChannel.some((v: number) => v < 0)
-    const length = leftChannel.length
+    ctx.fillStyle = this.convertColorValues(options.waveColor)
 
-    const barCount = Math.floor(width / (barWidth + barGap))
-    const barIndexScale = barCount / length
-    const halfHeight = height / 2
+    // Custom rendering function
+    if (options.renderFunction) {
+      options.renderFunction(channelData, ctx)
```

Comment:
Amazing, exactly what i needed in https://github.com/katspaugh/wavesurfer.js/discussions/2815

Reply:
Not sure I follow